Frameworks/CoreGraphics/CGDataProvider.mm

```diff
@@ -16,130 +16,395 @@
+        return (*(__CGDataProvider*)lhs == *(__CGDataProvider*)rhs);
+    }
+
+    static CFHashCode __CFHash(CFTypeRef provider) {
+        return ((__CGDataProvider*)provider)->m_internal->Hash();
+    }
+};
+
+#pragma endregion // __CGDataProvider
+
+#pragma region CGDataProvider Functions
 /**
  @Status Interoperable
+ @Notes
 */
-CFDataRef CGDataProviderCopyData(CGDataProviderRef provider) {
-    uint8_t* bytes = (uint8_t*)[(NSData*)provider bytes];
-    size_t length = [(NSData*)provider length];
-    return CFDataCreate(nullptr, bytes, length);
+CGDataProviderRef CGDataProviderCreate(void* info, const CGDataProviderCallbacks* callbacks) {
+    return __CGDataProvider::CreateInstance(nullptr, info, callbacks);
 }
 
 /**
- @Status Caveat
- @Notes Only calls getBytePointer
+ @Status Interoperable
+ @Notes
 */
-CGDataProviderRef CGDataProviderCreateDirect(void* info, off_t size, const CGDataProviderDirectCallbacks* callBacks) {
-    TraceWarning(TAG, L"Warning: CGDataProviderCreateDirect is hacky");
-    char* pBytes = (char*)callBacks->getBytePointer(info);
+CGDataProviderRef CGDataProviderCreateDirectAccess(void* info, size_t size, const CGDataProviderDirectAccessCallbacks* callbacks) {
+    return __CGDataProvider::CreateInstance(nullptr, info, size, callbacks);
+}
+/**
+ @Status Interoperable
+ @Notes
+*/
+CGDataProviderRef CGDataProviderCreateDirect(void* info, off_t size, const CGDataProviderDirectCallbacks* callbacks) {
+    return __CGDataProvider::CreateInstance(nullptr, info, size, callbacks);
+}
 
-    id ret = [[CGDataProvider alloc] initWithBytesNoCopy:pBytes length:(DWORD)size freeWhenDone:FALSE];
+/**
+ @Status Interoperable
+ @Notes
+*/
+CGDataProviderRef CGDataProviderCreateSequential(void* info, const CGDataProviderSequentialCallbacks* callbacks) {
+    return __CGDataProvider::CreateInstance(nullptr, info, callbacks);
+}
 
-    return ret;
+/**
+# @Status Interoperable
+*/
+CGDataProviderRef CGDataProviderCreateWithData(void* info, const void* data, size_t size, CGDataProviderReleaseDataCallback releaseData) {
+    return __CGDataProvider::CreateInstance(nullptr, info, data, size, releaseData);
 }
 
 /**
  @Status Caveat
- @Notes Hacky
+ @Notes Only supports file:/// URLs
 */
-CGDataProviderRef CGDataProviderCreateSequential(void* info, const CGDataProviderSequentialCallbacks* callBacks) {
-    TraceWarning(TAG, L"Warning: CGDataProviderCreateSequential is hacky");
-    char* pBytes = (char*)IwMalloc(1024 * 1024);
-
-    int amt = callBacks->getBytes(info, pBytes, 1024 * 1024);
-
-    id ret = [[CGDataProvider alloc] initWithBytes:pBytes length:amt];
-    IwFree(pBytes);
-
-    return ret;
+CGDataProviderRef CGDataProviderCreateWithURL(CFURLRef url) {
+    return __CGDataProvider::CreateInstance(nullptr, url);
 }
 
 /**
  @Status Interoperable
 */
 CGDataProviderRef CGDataProviderCreateWithCFData(CFDataRef data) {
-    return (CGDataProviderRef)CFRetain(data);
+    return __CGDataProvider::CreateInstance(nullptr, data);
```

Comment:
you aren't checking for invalid state here nor within the internal implementation.
please do so here and elsewhere